Postoffice: Use atomic conf file write

utopiabound opened the pull request:
Write a new file and rename over old.
The rename is atomic.

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/whamcloud/integrated-manager-for-lustre/2344)
<!-- Reviewable:end -->

utopiabound committed:
Postoffice: Use atomic conf file write

Write a new file and rename over old.
The rename is atomic.
Empty lines are invalid

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

jgrund commented:
I'm not quite sure what problem this is solving, but why not use an actual mutex singleton? That would ensure a single writer in an idiomatic way.

mkpankov commented on iml-agent/src/action_plugins/postoffice.rs:
```diff
@@ -2,43 +2,118 @@
+async fn locked_route(
+    file: fs::File,
+    mailbox: String,
+    insert: bool,
```
```
enum Operation {
    Insert,
    Remove
}
```
is probably easier to understand

johnsonw commented:
@jgrund, would using a combination of lazy static with a mutex work?

utopiabound commented:
The provides Reader/Writer protection without a mutex, and without worrying about ensuring all threads have access to a shared mutex.  A reader will ALWAYS get a well formatted complete list regardless of when they read, and will never have to block on a writer.  A writer will never leave a file half updated (since the tmp file is deleted at startup).

Commit:
Remove devel-release workflow

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

Commit:
Add iml-common and iml-agent in tree

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

Commit:
Nuke ZFS

EXAScaler does not use ZFS so we can remove all of the ZFS related code
from EMF.

If it turns out it's needed again in the future, we can reference this
PR and re-implement the Rust bits.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

Commit:
Move integration testing to github-actions (#9)

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

Commit:
fixup integration-testing.yml

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

Commit:
Update Integration Testing

  - Use a single workflow to build RPMs and test them.
  - Remove reliance on Jenkins for PRs.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

Commit:
Update docker setup (#2444)

Update the IML docker setup to not print output while starting.

In addition, use a container that becomes stable earlier to copy
settings to the host.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

beevans committed:
EX-1979 filesync restore (#2367)

* EX-1979 filesync restore

First pass at filesync restore

Signed-off-by: Ben Evans <beevans@whamcloud.com>

* Enable lipe scan for restore

Signed-off-by: Ben Evans <beevans@whamcloud.com>

* Fix issue with 'Push' vs. 'push'

Signed-off-by: Ben Evans <beevans@whamcloud.com>

* clippy fixes

Signed-off-by: Ben Evans <beevans@whamcloud.com>

* Fileysnc data mover rewrite

Signed-off-by: Ben Evans <beevans@whamcloud.com>

* Fixed formatting

Signed-off-by: Ben Evans <beevans@whamcloud.com>

* Added code for adding tasks and tasklists
No need to have a scan job run.

Signed-off-by: Ben Evans <beevans@whamcloud.com>

* added code to create a task independent of Lipe.

Signed-off-by: Ben Evans <beevans@whamcloud.com>

* juniper query improvements

Signed-off-by: Ben Evans <beevans@whamcloud.com>

* rebase

Signed-off-by: Ben Evans <beevans@whamcloud.com>

* Bump

Signed-off-by: Ben Evans <beevans@whamcloud.com>

* Functioning with push and pull, if the files are on lustre

Signed-off-by: Ben Evans <beevans@whamcloud.com>

* fix warnings, cargo cruft

Signed-off-by: Ben Evans <beevans@whamcloud.com>

* Fixes from review

Signed-off-by: Ben Evans <beevans@whamcloud.com>

* Get rid of two more unwraps

Signed-off-by: Ben Evans <beevans@whamcloud.com>

* Code review changes

Signed-off-by: Ben Evans <beevans@whamcloud.com>

* Use expression that is already pattern matched

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

Co-authored-by: Joe Grund <jgrund@whamcloud.io>

Commit:
Fixup copy-embedded-settings (#2448)

`copy-embedded-settings` will find the trailer before looping. If the
container dies after entering the loop, then the loop will never exit as
the trailer is then non-existent.

Instead, we should find the trailer each turn of the loop to ensure we
are referencing a live container.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

Commit:
Prune unused files (#14)

Prune some unused files. There is more to remove, but that can happen in
another pass.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

beevans committed:
EX-2304 filesync: create the right path (#2449)

use dest_dir instead of src_dir when creating
paths to copy data

Signed-off-by: Ben Evans <beevans@whamcloud.com>

Commit:
bump (#15)

* Bump deps

- Update juniper to released version on crates.io
- Update sqlx to released version on crates.io

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

* Updates

- Use tagged github version of sqlx-cli, as 0.3 is not released yet
- Bump all insta usages to latest 1.x.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

beevans committed:
Fix for issue 2446 (#2447)

* Fix for issue 2446

Cloudsync and filesync should support the same command lines

Signed-off-by: Ben Evans <beevans@whamcloud.com>

* fix clippy errors

Signed-off-by: Ben Evans <beevans@whamcloud.com>

Commit:
Fix client mount generation (#2450)

Latest version of EXAScaler appears to have a number of new resource
constraints that weren't in the previous version.

This in turn has exposed a bug where we are not filtering by banned resource properly
and are picking up all target bans, not just the MGS one.

Instead, we should match the resource mount_point with the target
mount_path to only include the relevant resource.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

beevans committed:
EX-2312 cloudsync: Fix cloudsync lipe config (#2451)

Cloudsync lipe config had a few filesync references in it

Signed-off-by: Ben Evans <beevans@whamcloud.com>

Commit:
Add test case for DDN-1731 (#16)

Add a test case for customer data found in DDN-1731.

Latest changes already support the issue (floating IP assigned to
high-speed network), this test ensures we don't regress if it happens
elsewhere.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

Commit:
Remove update-check (#26)

Remove iml-update-{check,handler} which was basically vestigial.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

utopiabound committed:
Do not try to check removed files (#35)

Skip removed files from format checking

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

Commit:
Remove Power Control service (#27)

We no longer need to monitor outlet states via telnet from the manager.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

utopiabound committed:
Add iml-sos-plugin RPM (#37)

* Add iml-sos-plugin RPM

Import iml_sos_plugin iml.py from master (6700d91)
Build only sos/plugin/iml.py, not seperate cli
Rename RPM to something sane.
Add to integration testing.
use setuptools to build rpm

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

* Fix integration build

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

* Rename iml-sos to emf-sos

Use spec file instead of python build rpm

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

* Fix unremoved artifact

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

* Update emf-sos-plugin/README.md

Co-authored-by: Joe Grund <jgrund@whamcloud.io>

* Update a few spots

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

Co-authored-by: Joe Grund <jgrund@whamcloud.io>

Commit:
Update builds (#36)

Update the rust-base.dockerfile and integration-testing.yml to reuse cache as much as possible.

This appears to shave 20 - 30 minutes off the integration-testing workflow in the general case (dependencies have not changed).

- Add a new timings github actions job that will report build times for each push
- Move all base containers to CentOS 8
- Remove rust bits from python-service-base container
- Use `cargo-chef` to try and improve rust-base build times in CI.
- Fix garbage output in `wait-for-dependencies-postgres.sh`
- Bump `iml-common` to 1.5 and use that version in the manager and agent.
- Silence verbose logging from `iml-device`
- Trim unused deps in the manager
- Regenerate `requirements.txt` using `pipenv`
- Update EMF sos plugin to collect some CLI calls.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

Commit:
Remove all copytool code. (#28)

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

Commit:
Fixup some python code (#29)

Remove some dead python code, and fix some imports that were not
resolvable by IDE.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

ip1981 committed:
Use strict and verbose shell in integration-testing.yml (#53)

Helps catch and understand problems

Signed-off-by: Igor Pashev <pashev.igor@gmail.com>

Commit:
Attempt to fix iml-common for local builders. (#52)

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

Commit:
Add file push to iml-ssh (#71)

Add a simple way to stream anything that implements `AsyncRead` to a
remote node over SSH.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

Commit:
Add workflow to remove old artifacts (#72)

Create a new workflow to manually remove old artifacts

Commit:
Refactor build system (#55)

* Refactor build system

Fixes #54

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

* Add in make clean / extraclean

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

* update integration-testing.yml

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

* take 2

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

* take 3

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

* take 4

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

* take 5

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

* take 6

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

Co-authored-by: Nathaniel Clark <nclark@whamcloud.com>

LouisDDN committed:
Cancel previous runs when git push (CI) (#80)

RDruon committed:
IML -> EMF package renaming (#75)

* Rename folder and filename

* Replace occurence of IML to EMF

utopiabound committed:
Update Vagrantfile to work with VBox better (#78)

* Update Vagrantfile to work with VBox better

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

* Apply suggestions from code review

Co-authored-by: Joe Grund <jgrund@whamcloud.io>

* Check for file is file

Start ports indexing at 0 instead of 1

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

* Account for actual VirtualBox name

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

Co-authored-by: Joe Grund <jgrund@whamcloud.io>

Commit:
Wire up nightly builds (#85)

Create a workflow that will run at midnight (or manually) and push
latest master artifacts to jenkins.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

utopiabound committed:
Add es-vagrant directory (#82)

* Add es-vagrant directory

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

* Address review comments

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

utopiabound committed:
Adjust hostname names to point to .10 IPs (#93)

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

utopiabound committed:
Fix Ubuntu Client build (#96)

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

utopiabound committed:
Add EXA "base" build to integration tests. (#100)

* Fix Ubuntu Client build

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

* Add EXA "base" build to integration tests.

Fixes #92

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

* No SimpleServer, do don't try to kill it

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

utopiabound committed:
es-vagrant: Reduce storage requirements (#106)

Remove unused OSTs and MDT
This roughly halves the on-disk usage of running es-vagrant

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

Commit:
Implement consul bootstrap CLI (#81)

* Implement consul CLI

In order to create a consul datacenter, it first needs to be
bootstrapped.

Create a CLI that will bootstrap a cluster.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

* Updates

- extract SSH config to separate module
- Cleanup CLI to make more user-friendly

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

mkpankov committed:
Install RPMs instead of attempting to update them (#108)

Signed-off-by: Michael Pankov <work@michaelpankov.com>

nlinker committed:
Bump Copyright to 2021 (#107)

* Bump Copyright to 2021

Signed-off-by: Nick Linker <nlinker@gmail.com>

* Typo fixed

Signed-off-by: Nick Linker <nlinker@gmail.com>

ip1981 committed:
Package agent for Ubuntu (#31)

Non-pure build (using cargo) and package

Update debian/changelog:

    DEBEMAIL='iml@whamcloud.com' DEBFULLNAME='EMF Team' dch ...

Build binary only (!) packages, from the tol-level directory:

    dpkg-buildpackage --build=binary [--no-sign]

Or:
    make deb
    make deb-repo

utopiabound committed:
es-vagrant Fix configure and teardown of esui (#116)

Fixes #97

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

RDruon committed:
Fix cargo chef (#117)

RDruon committed:
Add es-vagrant to .dockerignore (#118)

Since this folder can contains *.rpm

Commit:
fix sha

mkpankov committed:
Revert "Fix cargo chef (#117)" (#120)

This reverts commit daef3c43327395a14cd035fdce15120e89a24012.

Commit:
fix branch input

utopiabound committed:
Fix HA startup (#125)

Also fix OOM on reboot with small mem (1G) es nodes

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

utopiabound committed:
manager-cli: Add influx config (#110)

* manager-cli: Add influx config

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

* Address review comments

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

* Fix spelling

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

* Add colored help

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

Commit:
Fix incorrect provides and obsoletes (#138)

utopiabound committed:
manager-cli: Add postgres config (#133)

* manager-cli: Add postgres config

Related #101

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

* Add colored help to subcommand

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

* manager-cli: Break out psql() and su() to utils

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

* Split up postgres into GenerateConfig and Setup

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

* Address clippy warning

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

* Add Grafana and update with dotenv and env settings

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

* Update emf-manager-cli/src/postgres.rs

Co-authored-by: Joe Grund <jgrund@whamcloud.io>

Co-authored-by: Joe Grund <jgrund@whamcloud.io>

Commit:
Fixups to emf-config cli (#141)

When testing the emf-config cli I found a few things that need to be
fixed in order for deployment to work.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

ip1981 committed:
Fix spaces issue in make destroy (#146)

Signed-off-by: Igor Pashev <pashev.igor@gmail.com>

utopiabound committed:
EX-1907 es-vagrant: Add create/destroy emf (#147)

* EX-1907 es-vagrant: Add create/destroy emf

Also minor cleanups in Makefile
Utilize second MDT in filesystem

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

* Fix typo

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

* Address review comments

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

ip1981 committed:
State Machine Initialization: Schema Validation (#111)

Signed-off-by: Igor Pashev <pashev.igor@gmail.com>

mkpankov committed:
Fix server profiles fetching incorrectly (#148)

Signed-off-by: Michael Pankov <work@michaelpankov.com>

Commit:
Implement nginx config for self-signed certs (#144)

* Implement nginx config for self-signed certs

This patch adds a new `emf-config nginx` command, `generate-self-certs`
It can be used to bootstrap certs for nginx.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

* Address review comments

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

mkpankov committed:
Fix the Cargo warning (#150)

Signed-off-by: Michael Pankov <work@michaelpankov.com>

Commit:
Remove all python code (#109)

This patch removes all the python code in the project that is scheduled
to be replaced in the 6.0 release.

It is not intended to land standalone, but to be a base for 6.0 work.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

utopiabound committed:
Remove consul and fixup help messages (#152)

Ensure all helps are colorful
Ensure all helps list long and short options
Ensure all configs are settable via ENV variable

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

ip1981 committed:
Do not build unused SRPMs (#154)

Signed-off-by: Igor Pashev <pashev.igor@gmail.com>

RDruon committed:
Add missing make check and make fmt for emf-gui (#140)

RDruon committed:
Create exascaler.conf parser (#88)

* Initial implementation of EXAScaler struct

* Implement parser

* EXAScaler parser using Python API

* Update emf-exa-parser/Cargo.toml

Co-authored-by: Joe Grund <jgrund@whamcloud.io>

* Use tokio

* Remove serde rename

* Remove unwrap

* Remove new line in struct

* Add copyright and remove zip

* Add test where exascaler.conf is not valid

* ::serde_json::Value -> serde_json::Value

* Remove serde_json::Value

* Use thiserror crate and remove json cleanup (done in python)

* Fix test

* Fix zpool_settings

* Use integration tests instead of unit tests

* Add integration test to rust-ci

Co-authored-by: Joe Grund <jgrund@whamcloud.io>

RDruon committed:
Use non_negative_derivative to calculate the cpu percentage since /proc/stat report total time spent since last boot (#153)

Fix long_term query, current query did not report value older than a day

utopiabound committed:
EX-2410 emf-config: Add kuma (#156)

* EX-2410 emf-config: Add kuma

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

* Fix clippy warning

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

Commit:
Update EMF to use service-mesh pattern (#130)

* Update EMF to use service-mesh pattern

Update EMF to utilize the service-mesh pattern.

The EMF agent was previously a single daemon that hosted a number of
plugins. These plugins all ran within the same process and utilized
futures and tasks to run.

This patch updates the agents to run as standalone processes each
installed by a separate RPM.

Each process is associated with an envoy process for network IO, and
orchestrated by a kuma server.

This means each process only needs to know the ports it needs to
communicate with it's sidecar as all network IO from the process happens
over localhost.

This patch also moves all relevant database tables into sqlx as all
migration is done there.

Rabbitmq has been removed, it is replaced by direct comms between
services through envoy.

Docker has also been removed, the manager and agent can co-exist on the
same node, so it is no longer needed.

The old vagrant instance has been removed as we are focusing on
EXAScaler exclusively.

Finally, all job-scheduler command and action code has been removed from
the backend. It will be replaced by the state-machine work currently in
design.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

* address review comments

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

* Fix broken nginx snapshot test

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

* remove postoffice sidecar

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

Commit:
Add kuma rpm into release  (#157)

* Add kuma rpm into release

Add the kuma rpms from https://github.com/whamcloud/pkg-kuma into
generated rpm bundle.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

* Apply suggestions from code review

* Create dir before downloading kuma binaries.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

* Update .github/workflows/integration-testing.yml

* Update .github/workflows/integration-testing.yml

* update path for kuma

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

Commit:
Remove glob expansion when deleting old sockets (#164)

Replace glob expansion which does not work, with machine_id expansion.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

ip1981 committed:
Use strict bash shell in Makefiles (#162)

We are already using some bashisms in Makefiles, for example {a,b,c}.

Signed-off-by: Igor Pashev <pashev.igor@gmail.com>

Commit:
Update ubuntu packaging (#160)

* Update ubuntu packaging

Update ubuntu packaging to work with service-mesh

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

* Address review comments

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

utopiabound committed:
EX-2410 config: Add bootstrap (#163)

* EX-2410 config: Add bootstrap

Bootstrap local configuration

override of PGDATA dir done via systemd dropin

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

* Idempotency of config, and rpm packaging

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

* Address clippy warnings

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

* Fix typo

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

* Wait for influx

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

* Update emf-manager-cli/src/config-main.rs

Co-authored-by: Joe Grund <jgrund@whamcloud.io>

* Fix clippy warnings

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

Co-authored-by: Joe Grund <jgrund@whamcloud.io>

Commit:
Database: Create input_document table in postgres (#169)

* Database: Create input_document table in postgres

Since we are recording all of the input documents, we need a table to store them into the database. The table name could be something like, "input_document." This table should be added using a sqlx migration.

Signed-off-by: johnsonw <wjohnson@whamcloud.com>

* Default to 'now()'

Signed-off-by: johnsonw <wjohnson@whamcloud.com>

Commit:
remove enp0s9 gateway (#171)

Fixes #145

Commit:
Generate kuma certs (#168)

* Generate kuma certs

Kuma will generate certs by default that live under /tmp and are removed
on service restart. This in turn will cause all existing dataplane
services to no longer be able to auth with the control plane when it
comes back up and certs are regenerated.

Ensure the control plane has static certs that don't change during
reboot, and that the dataplanes use them.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

* CR updates

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

* CR updates

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

* pin download action version

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

Commit:
Bump kuma dep to 1.0.7 (#173)

* Generate kuma certs

Kuma will generate certs by default that live under /tmp and are removed
on service restart. This in turn will cause all existing dataplane
services to no longer be able to auth with the control plane when it
comes back up and certs are regenerated.

Ensure the control plane has static certs that don't change during
reboot, and that the dataplanes use them.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

* CR updates

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

* CR updates

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

* pin download action version

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

* Bump kuma dep to 1.0.7

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

utopiabound committed:
Move to Tokio 1.x (#175)

Also:
reqwest 0.11
sqlx 0.5
bytes 1
hyper 0.14
thrush 0.32

Changes of note:
* sqlx 0.5 requires any crate using sqlx macros to import sqlx directly,
and not via emf-postgres (this is now not possible)
* tokio 1
tokio::time::delay_for -> tokio::time::sleep
UnixListener, UnboundedReceiver and ReadDir need tokio-stream wrappers to be streams (see UnixListenerStream, UnboundedReceiverStream, and ReadDirStream)

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

Commit:
Run cargo fmt (#176)

run `cargo fmt` to update files that no longer pass with rust 1.50.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

Commit:
Fixup kuma bootstrap (#177)

- We should return early if certs already exist.
- We need to wait for kuma to become available before creating policies.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

Commit:
Handle send failures from agent services (#179)

If an agent service fails to send through it's sidecar, it currently
fails and then sits there forever (because we don't listen for errors in
`writer.send`).

Instead, we can not propagate the error and instead keep the cached
value as `None`.

This will ensure the next time we write it will attempt to create a new
connection to the sidecar.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

Commit:
EX-2430 - Initialization: Parse state-schema.yml into struct (#161)

* EX-2430 - Initialization: Parse state-schema.yml into struct

When the state machine loads, part of the initialization requires that
it will parse *state-schema.yml* into a struct that will live in memory.
This procedure must be complete before the state transition graph can be
created.

Signed-off-by: johnsonw <wjohnson@whamcloud.com>

* - Fix clippy error
- small updates to the state_schema for provisional items

Signed-off-by: johnsonw <wjohnson@whamcloud.com>

* - Remove visitor implementation and use serde's "try_from"

Signed-off-by: johnsonw <wjohnson@whamcloud.com>

ip1981 committed:
Disable debug packages (#184)

Signed-off-by: Igor Pashev <pashev.igor@gmail.com>

Commit:
Implement staleness check (#183)

If a manager service dies, it needs to signal it's corresponding agent so data can be re-sent.

An efficient way to do this is to simply boot with a uuid and expose it via an HTTP endpoint.

Agents can periodically poll this endpoint and resend data iff the guid has changed since a previous poll.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

RDruon committed:
Fix systemd path (#191)

utopiabound committed:
agent: Fix clippy warnings (#189)

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

utopiabound committed:
config: Fix whitespace output when configuring influx (#192)

Fix extra newlines

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

Commit:
Bump envoy version (#187)

Kuma 1.0.7 ships with a newer version of envoy. Bump our version to one
slightly newer as the matching version was yanked from the envoy repos.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

utopiabound committed:
agent: Remove extra call to crm_mon in get_crm_mon() (#188)

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

RDruon committed:
Fix EX-2582 (#198)

Commit:
Use full SHA for actions (#200)

Github actions now require the full SHA when pinning actions.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

RDruon committed:
Fix EX-2581 (#199)

* Fix EX-2581

* Fix format and better code handling

* Update device-scanner/mount-emitter/src/bin/swap_emitter.rs

Co-authored-by: Joe Grund <jgrund@whamcloud.io>

Co-authored-by: Joe Grund <jgrund@whamcloud.io>

Commit:
EX-2432 - Initialization: State Transition Graph (#167)

* Initialization: State Transition Graph

When the state machine first loads it will load a schema describing all components, their corresponding states, and actions. As part of the initialization task, the state machine will need to parse the schema and build a graph representing all possible state transitions for components.

Signed-off-by: johnsonw <wjohnson@whamcloud.com>

* - Add a "TransitionGraph" type

Signed-off-by: johnsonw <wjohnson@whamcloud.com>

* - Updates after rebasing

Signed-off-by: johnsonw <wjohnson@whamcloud.com>

* - Remove the ComponentType struct from the state machine crate and
instead extend the existing struct in the wire-types crate.

Signed-off-by: johnsonw <wjohnson@whamcloud.com>

* - CR updates

Signed-off-by: johnsonw <wjohnson@whamcloud.com>

* - More updates

Signed-off-by: johnsonw <wjohnson@whamcloud.com>

* - Remove unnecessary import

Signed-off-by: johnsonw <wjohnson@whamcloud.com>

* - more updates

Signed-off-by: johnsonw <wjohnson@whamcloud.com>

* - Fix ntp component in warp-drive

Signed-off-by: johnsonw <wjohnson@whamcloud.com>

* - fix formatting

Signed-off-by: johnsonw <wjohnson@whamcloud.com>

Commit:
EX-2427 - Input Document - Input Handler - Input Document Parser (#172)

* Input Document - Input Handler - Input Document Parser

Parse an input document into an InputDocument structure.

Signed-off-by: johnsonw <wjohnson@whamcloud.com>

* CR updates

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

* More CR updates

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

Co-authored-by: Joe Grund <jgrund@whamcloud.io>

RDruon committed:
Fix EX-2583 (#205)

* Add kuma-dp download to intergration-testing

* Fix kuma dp addition to repo

Commit:
Fixup sqlx data (#207)

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

Commit:
Switch from unimplemented! to todo! (#208)

Switch from `unimplemented!()` to `todo!()`, which signals intent
better.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

Commit:
EX-2601: Input Document - Schema Validation (#174)

* EX-2601: Input Document - Schema Validation

When receiving an input document the schema for the document (in yaml format) must be validated. Write a function to validate the document against the schema.

Signed-off-by: johnsonw <wjohnson@whamcloud.com>

* - Fixup test

Signed-off-by: johnsonw <wjohnson@whamcloud.com>

* CR updates

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

* Fix formatting

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

Co-authored-by: Joe Grund <jgrund@whamcloud.io>

RDruon committed:
Fix EX-2584 (#206)

* Add envoy packaging to ubuntu integration testing

* Use curl to download envoy, fix envoy addition to repo

Co-authored-by: Joe Grund <jgrund@whamcloud.io>

nlinker committed:
Fix policy (#210)

Signed-off-by: Nick Linker <nlinker@gmail.com>

ip1981 committed:
Use Kuma 1.0.8 (#215)

Signed-off-by: Igor Pashev <pashev.igor@gmail.com>

Commit:
Implement static schema and precise validation (#213)

* Implement static schema and precise validation

The state-schema is currently a .yml file that is both parsed and loaded
at the start of runtime.

After thinking about this a bit, there is no reason why the state-schema
needs to be implemented in this way. The state-schema is only
meant for developer usage.

Instead, we can define the state-schema in rust, which will allow us to
catch any errors at compile time instead of runtime / in unit tests.

In addition, we can use rust for validation of input documents as well,
rather than using jsonschema. This should allow for a better user
experience.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>
Signed-off-by: johnsonw <wjohnson@whamcloud.com>

* - Update StateSchema structure

Signed-off-by: johnsonw <wjohnson@whamcloud.com>

* Make TransitionGraph types ord just so tests are stable.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

* - Make "tags" optional in client_mount Create input
- Add MgtType enum to filesystem to ensure only the appropriate value
can be passed in
- Add input for "unmount" action in mgt
- Add remaining action match arms
- Add tests for each component
- Add some validation tests. There are still a few more I believe I can
add in a follow up patch.

Signed-off-by: johnsonw <wjohnson@whamcloud.com>

* Remove unused code and format

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

* - Testing updates

Signed-off-by: johnsonw <wjohnson@whamcloud.com>

* - fix clippy errors

Signed-off-by: johnsonw <wjohnson@whamcloud.com>

* Remove unused schemas

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

Co-authored-by: johnsonw <wjohnson@whamcloud.com>

RDruon committed:
Fixes EX-2586 (Add apt_repo endpoint to emf gateway service) (#212)

* Serve rpm and deb through nginx gateway

* Move repo rpm creation from Makefile to integration-testing

* Fix format error

* Apply suggestions from code review

Co-authored-by: Joe Grund <jgrund@whamcloud.io>

Co-authored-by: Joe Grund <jgrund@whamcloud.io>

Commit:
Implement action-runner services (#203)

* Implement action-runner services

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

* fixup broken tests

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

* fixup broken example

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

* Fix example

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

* fixup spec

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

Commit:
EX-2480 - API - Create endpoint to receive input documents (#196)

* EX-2480 - API - Create endpoint to receive input documents

The state machine must have a way to receive input documents. These
documents can be submitted to the state machine by using a graphql
endpoint. When the call is received, two things should happen:

1. Record the submitted input document in jsonb format into the
"input_document" table if validation passes
2. Submit the input document to the state machine for processing

After the document is submitted, the state machine should return a
Command object. This part will be handled in a separate ticket.

Signed-off-by: johnsonw <wjohnson@whamcloud.com>
Signed-off-by: Joe Grund <jgrund@whamcloud.io>

* update sqlx-data

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

Co-authored-by: Joe Grund <jgrund@whamcloud.io>

utopiabound committed:
Build sos rpm in all target (#219)

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

RDruon committed:
Fixes EX-2585 (Ubuntu ntp service updates) (#216)

* Look for ntpd.service or ntp.service

* Use service file as parameter

* Remove comment

* Rename F[5-6] to F[3-4]

* Merge ntp_runstate and ntp_ubuntu_runstate

* Fix review comments

* Merge new types

* Fix review comments

* Fix review comment on test

* Fix review comments. Relies on file existence instead of hiding errors

Commit:
Implement first pass of execution pipeline (#218)

* Implement first passs of execution pipeline

Implement a first incomplete pass of taking an input document and doing
a kuma dp deployment.

This is very incomplete, and will be fleshed out further in this and
other patches.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

* Updates

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

* Various fixups

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

* fixup ssh calls

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

* - Add graph execution stack population test
- Change "cmd" to "run" in deploy-host script

Signed-off-by: johnsonw <wjohnson@whamcloud.com>

* Some cleanups

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

* More fixups

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

* More fixups

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

* Fixup tests

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

* More fixups

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

* Propagate failed or cancelled steps to all children

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

* Update emf-state-machine/src/command_plan.rs

Co-authored-by: johnsonw <wjohnson@whamcloud.com>

Commit:
EX-2301: Fixup error message when cli config missing (#221)

Fixes [EX-2301](https://jira.whamcloud.com/browse/EX-2301)

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

Commit:
Increase default MGT size to accomodate VG restrictions (#222)

EX-2395 introduced a restriction on the emf lv to be at least 80G.

Update the MGS storage to accomodate this new restriction.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

Commit:
Implement run cli (#223)

Implement a new cli that can take a path to an input document and send it
to the graphql api.

In the future, this cli will be expanded to block on command completion,
for now it just returns as soon as the input document has been
submitted.

```
emf run -h`
emf-run 0.4.0
Run an input document on the EMF state-machine

USAGE:
    emf run <path>

FLAGS:
    -h, --help       Prints help information
    -V, --version    Prints version information

ARGS:
    <path>    Path of input document
```

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

Commit:
Vendor all RPM deps (#220)

Vendor all RPM based deps. This should ease integration with vendoring
in EXAScaler 6.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

Update integration-testing.yml
Update .github/workflows/integration-testing.yml

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

Commit:
Propagate input-document parse errors back to client (#224)

Update graphql server error handling to properly return upstream errors to the
client.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

Commit:
Fix cache invalidation (#225)

The cache invalidation was written in such a way that the server's
instance_id was never fetched.

Update the initial value so it works.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

Commit:
Implement action to create file on remote node via ssh (#226)

Implement a new action to create a file on a remote node via SSH.

This can be used to deploy ubuntu to dgx systems:

```yml
version: 1
jobs:
  deploy_ubuntu1:
    name: Deploy Ubuntu1

    steps:
      - action: host.setup_planes_ssh
        id: Setup Control plane and dataplane
        inputs:
          hosts:
            - ubuntu1
          cp_addr: node1

      - action: host.sync_file_ssh
        id: Sync dataplane token across cluster
        inputs:
          from: /etc/emf/tokens/dataplane-token
          hosts:
            - ubuntu1

      - action: host.create_file_ssh
        id: add emf repo
        inputs:
          host: ubuntu1
          contents: deb [trusted=yes] http://node1:7080/apt-repo/ emf non-free
          path: /etc/apt/sources.list.d/emf.list

      - action: host.ssh_command
        id: Install emf agent ubuntu1
        inputs:
          host: ubuntu1
          run: |
            apt update
            apt install -y \
              emf-action-agent \
              emf-device-agent \
              emf-journal-agent \
              emf-network-agent \
              emf-host-agent \
              emf-ntp-agent \
              emf-stats-agent

      - action: host.ssh_command
        id: Start agent ubuntu1
        inputs:
          host: ubuntu1
          run: |
            systemctl enable --now \
            emf-action-agent \
            emf-device-agent \
            emf-journal-agent \
            emf-network-agent \
            emf-host-agent \
            emf-ntp-agent \
            emf-stats-agent \
            emf-agent.target
```

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

Commit:
Reload daemon post install / upgrade (#227)

Fixes #202.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

utopiabound committed:
github: Ensure all features are tested in cargo test (#233)

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

Commit:
Remove hostlist expansion within actions (#231)

* Remove hostlist expansion within actions

Each action needs the invariant that it operates over a single entity
for tracking purposes. Expansion should be done at a higher level,
both through some syntax in input documents and programatically building
via a cli interface.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

* Fix tests

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

* Remove unused imports

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

Commit:
Comment out gui dep (#232)

* Comment out gui dep

We are not (currently) building the GUI, remove it as a dep until we are
again.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

* remove dep from mkdir -p /var/folders/p1/2wqz07s14td5gb5mf9272bq8r4bkzw/T/tmp.TqiObHz7/release/rust-core
cargo build --release

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

Commit:
Implement command plan persistence

Fixes [EX-2786](https://jira.whamcloud.com/browse/EX-2786)

Implement persistence of running command plans.
This is implemented as a main db receiver and a series of scoped writers
so many updates to the plan can be queued in parallel.

In addition, implement return of a serialized command plan and rendering
in the CLI.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

RDruon committed:
Fixes EX-2589 (jobstats metrics collection) (#217)

* Initial implementation of jobstat collection

* Add continuous query for new measurement 'jobstats'

* Fix format

* JobstatMdt implementation

* Bump lustre_collector version

* Fix review comments

Commit:
Introduce graphql_json scalar (#237)

* Introduce graphql_json scalar

Introduce the `graphql_json` scalar lifted from @nlinker
https://github.com/nlinker/rust-graphql-json example repo

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

* CR updates

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

utopiabound committed:
es-vagrant: minor cleanups (#242)

Remove extra file copies
Add config ost pools provisioner
fix vagrant rsync in Makefile

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

utopiabound committed:
liblustreapi: Update bindgen version (#195)

Update libloading version in llapi

Also remove added linker.

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

utopiabound committed:
Update pacemaker wire-types (#230)

* Update pacemaker wire-types

Add tests for processing XML to wire-types
Add tests for XML to wire-types to crm shell
Convert agent HA action plugin to use new wire types.

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

* Add Constraint and template tests

Fixup missing items

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

* Apply suggestions from code review

Co-authored-by: Joe Grund <jgrund@whamcloud.io>

* Add Error type to pacemaker

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

* Fix wire-types cargo deps

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

Co-authored-by: Joe Grund <jgrund@whamcloud.io>

Commit:
EX-2731 - Add spawn function to emf-ssh (#240)

* EX-2731 - Add spawn function to emf-ssh

Signed-off-by: johnsonw <wjohnson@whamcloud.com>

* Fixups

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

Co-authored-by: Joe Grund <jgrund@whamcloud.io>

Commit:
Add job dependency support (#243)

All jobs submitted with an input document run in parallel by default.
However, there may be cases where jobs should depend on other jobs
completing. To support this usecase, implement `job.needs`, which is an
array of job_names to wait on for completion.

Internally, this is implemented by storing a top-level graph of jobs,
whose nodes are graphs of steps.

We do a topological sort over the job nodes to ensure we build up stacks
of futures without missing any dependencies.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

Commit:
Fix SSH spawn output collection (#250)

When running commands, there may be actions that randomly have no stdout
collected when there should be.

This is because we currently break when we receive ExitStatus in spawn,
and it could be the case that data is written after that.

Instead of breaking when we get ExitStatus, just let `wait` run itself
out until it returns none.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

Commit:
Fix emf-ssh spawn tests (#247)

* - The spawn integration tests were failing due to the tests starting a
server on the same port in parallel. I've updated the tests to run on
separate ports so that they can still run in parallel but not overlap
each other.

Signed-off-by: johnsonw <wjohnson@whamcloud.com>

* - The stderr message is different depending on which operating system
the test is being run on. We just need to make sure that the buffer
contains "stderr: ls:" and that should satisfy the test regardless of
the operating system.

Signed-off-by: johnsonw <wjohnson@whamcloud.com>

* - The stderr test returns an exit code of 1 on OS X and an exit code of
2 on linux. Account for both.

Signed-off-by: johnsonw <wjohnson@whamcloud.com>

RDruon committed:
Add missing bootstrap.conf in state machine service (#256)

Commit:
Implement command detail CLIs (#252)

* Implement command detail CLIs

Implement CLIs to display commands at a summary (tree) view and a
detailed view.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

* Remove color from testing and remove some unused deps

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

Commit:
Implement command listing CLI (#253)

* Implement command listing CLI

Implement a simple CLI to list commands. It is missing any sort of
filtering and Name of the command, both of which can be added in
separate passes.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

* Sort ids

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

Commit:
Add new action to wait for host availibility (#257)

Add a new action that will wait for the given host to report in via the
EMF service-mesh.

It can be added to the end of host deployment to verify that the host is
available and that we can start issuing remote action-plugins.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

Commit:
Add SSH proxy support (#264)

When configuring EXAScaler at an early stage, we may not have reliable
networking. Because of this, we cannot do direct SSH and need to hop
through a controller prior to reaching a VM.

In order to do this, we need to add the ability to proxy commands
through a server, similar to a jump host or `ProxyCommand`.

Add new functionality to do this to emf-ssh.
Proxying can be done either via pre-configured key, or via password.

In addition, ensure all actions are able to handle proxying in their
inputs.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

utopiabound committed:
Rework exa-parser to simplify (#263)

* Rework exa-parser to simplify

Use es_config_show --api for json output
Unify all the config structs into a single file
Order the structs so values are more redily findable
Fix tests to just test serde parsing of es_config_show output

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

* Fix sorting

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

* Address review comments

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

utopiabound committed:
exa-parser: Host struct fixes (#271)

Ensure Nics is public
Remove erroneous Option from inside Vec in SfaSettings
Fix erroneous Nics in Hosts
Make singular structs singular

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

utopiabound committed:
exa-parser: Account for bonded interfaces (#274)

Also the es7k_with_extMDS exascaler.conf showed a number of other
issues in the structs.

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

Commit:
Add Configure IfConfig task to Host input (#265)

* Add Configure IfConfig task to Host input

Add Configure IfConfig task to Host input

Signed-off-by: johnsonw <wjohnson@whamcloud.com>

* Rebase and include proxy info when creating the session.

Signed-off-by: johnsonw <wjohnson@whamcloud.com>

* - Update to run ifconfig for a single nic, not a map of nics.

Signed-off-by: johnsonw <wjohnson@whamcloud.com>

* Some updates

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

* - Derive default for AuthOpts
- Update action to use a vec instead of a string and then join on a
newline when ready to send.

Signed-off-by: johnsonw <wjohnson@whamcloud.com>

* - Just use default on the agent

Signed-off-by: johnsonw <wjohnson@whamcloud.com>

Co-authored-by: Joe Grund <jgrund@whamcloud.io>

utopiabound committed:
agent: Update action check_stonith to use new wire types (#244)

Update action plugin check_stonith() to use new pacemaker
wire-types

Add traits for instance attributes, meta attributes, and utilization.

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

Commit:
Update Readme (#276)

- Remove git version from `sqlx-cli`
- Document db migration cmd.

Commit:
Add Result to command detail output (#277)

If a command fails prior to any stdout / stderr being written it is
unclear what exactly went wrong from the command detail.

Add a new field that when populated will the display a error / ok
message set at the completion of the action.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

utopiabound committed:
EX-2395 es-vagrant: Fix drive failover (#246)

Use shareable drives instead of multiattach.
multiattach are actually differenced from the image, and not shared storage.
shareable drives require fully allocated "fixed" size drives.
This relies on EX-2395 patch to add workaround to es-install for 5.2.2-r259+
For post 5.2.2-r259 size should be "5G_FORCE"
For anything prior prior to 5.2.2-r259 it should be "5G"
For 5.2.2-r259 EMF should be disabled.

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

Commit:
Add ConfigureNetworkSsh Action (#275)

* Add ConfigureNetworkSsh Action

Add ConfigureNetworkSsh Action. This will update /etc/sysconfig/network
on the host machine, based on the gateway device.

Signed-off-by: johnsonw <wjohnson@whamcloud.com>

* - Change `Chmod` to `ChmodSsh`
- Cleanups to `ConfigureNetwork`

Signed-off-by: johnsonw <wjohnson@whamcloud.com>

* fixups

Signed-off-by: johnsonw <wjohnson@whamcloud.com>

utopiabound committed:
exa-parser: Fix insta version (#279)

Ensure common version of insta used thoughout project.

Signed-off-by: Nathaniel Clark <nclark@whamcloud.com>

Commit:
Implement Alert API / CLI (#282)

Implement a graphql query and CLI for listing alerts. This should also
be used to render Alerts in the GUI.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

Commit:
Put step into running state (#284)

When a step starts to run we need to put it into the `Running` state so
it is clear when viewing the command.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

Commit:
Use EXA 5.2.2_r2 image (#287)

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

Commit:
Extract lib-state-machine (#288)

Extract the pieces to build input documents via Rust to a separate
crate.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

Commit:
Fix SSH proxy (#292)

An erroneous command was being passed when we build the SSH proxy
command. Fix it.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

Commit:
Set LimitNOFILE for all sidecar services (#291)

Fixes #185.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

Commit:
Add make commands for rust-core-rpms with timestamp (#285)

Add the following make commands:

1. `rust-core-rpms-local`: Builds only the core rust rpms with a
   timestamp
1. `rust-core-rpms-docker-local`: Builds only the core rust rpms in
   docker with a timestamp + incremental building support via sccache.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

Commit:
Fix job status (#280)

* Fix job status

The job status is showing "pending" in some cases. Fix the logic to fold
over the state being "Completed" and taking the minimum value.

Signed-off-by: johnsonw <wjohnson@whamcloud.com>

* - Don't use "Pending" at the top level. Instead, we should use
  "Running".

Signed-off-by: johnsonw <wjohnson@whamcloud.com>

* - Fixups

Signed-off-by: johnsonw <wjohnson@whamcloud.com>

* - Use `Pending` for all pending and completed.

Signed-off-by: johnsonw <wjohnson@whamcloud.com>

* - Remove unnecessary mut

Signed-off-by: johnsonw <wjohnson@whamcloud.com>

* - cleanups
- Add SleepActionSsh

Signed-off-by: johnsonw <wjohnson@whamcloud.com>

* - Update the upper level command state to use `get_state` when
  retrieving the state for each graph.

Signed-off-by: johnsonw <wjohnson@whamcloud.com>

* - Remove sleep action

Signed-off-by: johnsonw <wjohnson@whamcloud.com>

Commit:
Fix ssh-server in integration tests (#272)

The ssh server was not sending eof or closing connections when a command
completed. This prevented the creation of new channels from the same
handle. This patch updates the server to send eof and close the channel
at the end of an exec_request.

Signed-off-by: johnsonw <wjohnson@whamcloud.com>

Commit:
Add host deployment API and CLI (#290)

- Add a new host mutation that can deploy n hosts in parallel using the
EMF state-machine.
- Add a new CLI to deploy hosts in parallel
- Rename the server CLI to host

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

Commit:
Bump Kuma and Envoy to latest  (#296)

* Bump Kuma and Envoy to latest

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

* DEB: use variables for kuma and envoy versions

Signed-off-by: Igor Pashev <pashev.igor@gmail.com>

Co-authored-by: Igor Pashev <pashev.igor@gmail.com>

RDruon committed:
Fix systemctl service uninstallation (#294)

* Add systemd_preun macro to uninstall systemctl service file for agent

* Add missing service removal

Commit:
Add ConfigureBonding action (#278)

* Add ConfigureBonding action

Add ConfigureBonding action.

Signed-off-by: johnsonw <wjohnson@whamcloud.com>

* - Remove the is_bonded field. It seems like we shouldn't even create a
section for this if it isn't bonded to begin with.

Signed-off-by: johnsonw <wjohnson@whamcloud.com>

* - update comment

Signed-off-by: johnsonw <wjohnson@whamcloud.com>

* - CR updates

Signed-off-by: johnsonw <wjohnson@whamcloud.com>

* - Remove unused code

Signed-off-by: johnsonw <wjohnson@whamcloud.com>

Commit:
Update services (#298)

- Enable agent sidecars when service is enabled
- Drop PartOf service for agent sidecars

Fixes #297.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

Commit:
Use the machine id instead of fqdn when checking if a host is available. (#302)

Use the machine id instead of fqdn when checking if a host is available.

Signed-off-by: johnsonw <wjohnson@whamcloud.com>

Commit:
Add overrides.conf for local node (#299)

Add the overrides.conf for the local node so we don't need to enter it
by hand.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

Commit:
Machine (#304)

* - Wait on host to be available by machine-id

Signed-off-by: johnsonw <wjohnson@whamcloud.com>

* - trim the machine_id

Signed-off-by: johnsonw <wjohnson@whamcloud.com>

LouisDDN committed:
Improve documentation (#301)

* Improve vagrant documentation

* Autoremove emf on all nodes

* Add how to ssh in README

* Add snapshot cmds in README

* Improve postgresql instructions

* Add how to upgrade VMs in README

* Fix rustup default nightly

* Fix typo in vagrant script

* Apply suggestions from code review

Co-authored-by: Joe Grund <jgrund@whamcloud.io>

* Update es-vagrant/scripts/emf-setup.sh

Co-authored-by: Joe Grund <jgrund@whamcloud.io>

* Add warning about lustre snapshot targets

* Split emf-setup into multiple provisioners

- emf-deploy (only node1)
- emf-install-repo (only node1)
- emf-remove-old-bits (all nodes)

* Add script to setup ~/.ssh/config automatically

* Improve update_ssh_config

* Remove x86_64 in emf_repo_rpm (CI)

Signed-off-by: ldouriez <ldouriez@ddn.com>

* Chore

- Fix msg in update_ssh
- Add DB creation in README
- Chore in README

Signed-off-by: ldouriez <ldouriez@ddn.com>

* Update es-vagrant/scripts/emf-remove-old-bits.sh

Co-authored-by: Joe Grund <jgrund@whamcloud.io>

* Update es-vagrant/scripts/emf-deploy.sh

Co-authored-by: Joe Grund <jgrund@whamcloud.io>

* Update README.md

Co-authored-by: Joe Grund <jgrund@whamcloud.io>

* Update README.md

Co-authored-by: Joe Grund <jgrund@whamcloud.io>

* Update README.md

Co-authored-by: Joe Grund <jgrund@whamcloud.io>

* Remove upgrade rpm section

Co-authored-by: Joe Grund <jgrund@whamcloud.io>

RDruon committed:
Initial implementation of IB monitoring (#166)

* Initial implementation of IB monitoring

* Fix format

* Fix deb creation

* Update ib agent code

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

* Fix path

* Fix merge

* Update after rebase

* Use continuous query for infiniband throughput

* Bum envoy version

* Fix review

* Send empty data in case of unavailable IB metrics and log it

* Bump kuma/envoy version

* Add service removal for IB agent

* Remove Result return type since no error are returned

* Use ubuntu_dgx flavor

Co-authored-by: Joe Grund <jgrund@whamcloud.io>

Commit:
Implement health cli (#314)

Implement a new `emf health` cli that can be used to get health status
of EMF.

For now we do a simple check of the manager services via kuma. This will
be expanded in the future.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

Commit:
The es_config_show command will report "No such file or directory" if (#315)

the absolute path is not used. Include the full path in the command.

Signed-off-by: johnsonw <wjohnson@whamcloud.com>

mkpankov committed:
Prevent wiping out `target` when updating sqlx data (#305)

Signed-off-by: Michael Pankov <work@michaelpankov.com>
